### PR TITLE
DetectionModel impl

### PR DESCRIPTION
**Pull request by dkurt**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under Apache 2 License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/dnn/include/opencv2/dnn/dnn.hpp [hunk 1]
```diff
@@ -1307,6 +1307,9 @@ CV__DNN_INLINE_NS_BEGIN
          CV_WRAP void detect(InputArray frame, CV_OUT std::vector<int>& classIds,
                              CV_OUT std::vector<float>& confidences, CV_OUT std::vector<Rect>& boxes,
                              float confThreshold = 0.5f, float nmsThreshold = 0.0f);
+     protected:
+         struct DetectionModel_Impl;
+         Ptr<DetectionModel_Impl> impl;
```
Comment (alalek) [inline modules/dnn/include/opencv2/dnn/dnn.hpp]: There is already `Impl` in `Model` class.
No need to keep 2 pointers on "implementations", re-use single one "derived" instance.

Comment (dkurt) [reply modules/dnn/include/opencv2/dnn/dnn.hpp]: Replaced to derived Impl usage. Please take a look

File: modules/dnn/src/model.cpp [hunk 2]
```diff
@@ -320,28 +320,41 @@ void SegmentationModel::segment(InputArray frame, OutputArray mask)
...
+                layer->nmsThreshold = 0;
+            }
         }
     }
-}
+};
 
 DetectionModel::DetectionModel(const String& model, const String& config)
-    : Model(model, config)
+    : Model()
```
Comment (alalek) [inline modules/dnn/src/model.cpp]: C++11 allows to call other ctor:
```
: DetectionModel(readNet(model, config))
{
    // nothing
}
```

File: modules/dnn/src/model.cpp [hunk 3]
```diff
@@ -320,28 +320,41 @@ void SegmentationModel::segment(InputArray frame, OutputArray mask)
...
 {
-    disableRegionNMS(getNetwork_());  // FIXIT Move to DetectionModel::Impl::initNet()
+    impl = makePtr<DetectionModel_Impl>();
+    impl->initNet(readNet(model, config));
+    impl.dynamicCast<DetectionModel_Impl>()->disableRegionNMS(getNetwork_());  // FIXIT Move to DetectionModel::Impl::initNet()
```
Comment (alalek) [inline modules/dnn/src/model.cpp]: > `/*virtual*/ void initNet()`

You may want to uncomment `virtual` there (it is internal implementation, so it is allowed)

**Review by alalek — COMMENTED** [review]
Looks good to me :+1:

**Pull request closed without merging**